src/core/components/pin.js

```diff
@@ -0,0 +1,589 @@
+          }
+        }
+      }
+      hashes.forEach((hash) => {
+        if (typeof hash === 'string') {
+          // example: '/ipfs/QmRootHash/links/by/name'
+          const matched = hash.match(/^(?:\/ipfs\/)?([^\/]+(?:\/[^\/]+)*)\/?$/)
+          if (!matched) {
+            return callback(new Error('invalid ipfs ref path'))
+          }
+          const split = matched[1].split('/')
+          const rootHash = multihashes.fromB58String(split[0])
+          const links = split.slice(1, split.length)
+          if (!links.length) {
+            normalized.update(rootHash)
+          } else {
+            // recursively follow named links to the target
+            const pathFn = (err, obj) => {
+              if (err) { return callback(err) }
+              if (links.length) {
+                const linkName = links.shift()
+                const nextLink = obj.links.filter((link) => {
+                  return (link.name === linkName)
+                })
+                if (!nextLink.length) {
+                  return callback(new Error(
+                    `no link named ${linkName} under ${obj.toJSON().Hash}`
+                  ))
+                }
+                const nextHash = nextLink[0].hash
+                self.object.get(nextHash, pathFn)
+              } else {
+                normalized.update(obj.multihash)
+              }
+            }
+            self.object.get(rootHash, pathFn)
+          }
+        } else {
+          normalized.update(hash)
+        }
+      })
+    }
+    catch (ex) {
+      return callback(ex)
+    }
+  }
+
+  function getRecursive (multihash, callback) {
+    // gets flat array of all DAGNodes in tree given by multihash
+    // (should this be part of dag.js API? it was in ipfs-merkle-dag)
+    dag.get(new CID(multihash), (err, res) => {
+      if (err) { return callback(err) }
+      const links = res.value.links
+      // leaf case
+      if (!links.length) {
+        return callback(null, [res.value])
+      }
+      // branch case
+      let nodes = []
+      links.forEach(link => {
+        getRecursive(link.multihash, (err, subNodes) => {
+          if (err) { return callback(err) }
+          nodes.push(subNodes)
+          if (nodes.length === links.length) {
+            return callback(null, _.flattenDeep(nodes))
+          }
+        })
+      })
+    })
+  }
+
+  const pin = {
+    types: {
+      direct: 'direct',
+      recursive: 'recursive',
+      indirect: 'indirect',
+      internal: 'internal',
+      all: 'all'
+    },
+
+    clear: () => {
+      directPins.clear()
+      recursivePins.clear()
+      internalPins.clear()
+    },
+
+    set: pinSet(dag),
+
+    add: promisify((hashes, options, callback) => {
+      // callback (err, pinset)
+      let recursive = true  // default
+      if (typeof options === 'function') {
+        callback = options
+      }
+      else if (options && options.recursive === false) {
+        recursive = false
+      }
```

Comment:
Line 157 to 163 can be replaced by

`const recursive = options.recursive === false ? false : true`